doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -90,11 +93,69 @@ Rewriting with Leibniz and setoid equality
      any of them can be rewritten.  If not specified, only the first occurrence
      in the conclusion is replaced.
 
-     If :n:`at @occs_nums` is specified, rewriting is always done with
-     :ref:`setoid rewriting <generalizedrewriting>`, even for Leibniz’s equality,
-     which means that you must `Require Setoid` to use that form.
-     However, note that :tacn:`rewrite` (even when using setoid rewriting) and
-     :tacn:`setoid_rewrite` don't behave identically (as already mentioned above).
+     .. note::
+
+        If :n:`at @occs_nums` is specified, rewriting is always done
+        with :ref:`setoid rewriting <generalizedrewriting>`, even for
+        Leibniz equality, which means that you must `Require
+        Setoid` to use that form.  However, note that :tacn:`rewrite`
+        (even when using setoid rewriting) and :tacn:`setoid_rewrite`
+        don't behave identically (as already mentioned above).
+
+     .. note::
+
+        For each selected hypothesis and/or the conclusion,
+        :tacn:`rewrite` finds the first matching subterm in depth-first search
+        order and binds the free variables of the lemma.  Then, if the `at` clause
+        is specified, only subterms with the same bindings for the free variables
+        will be counted in determining the occurrence number.  This limits
+        which matches of the lemma can be rewritten.  You can get around
+        this limitation by explicitly specifying how some or all of the free
+        variables are bound
+        by using a `with` clause (see :n:`@one_term_with_bindings`).
+        You can also use :tacn:`setoid_rewrite`, which does the matching before
+        binding the free variables of the lemma.
```

Comment:
I am a bit reluctant to mention `setoid_rewrite` here. It is a completely different beast. It is not just a matter of counting occurrences. First, it selects the other subterms modulo convertibility rather than identity. Second, it silently ignores `at` clauses, if they are part of a larger occurrence or if they do not occur. Third, it applies the lemma as many times as there are occurrences, despite them being convertible. So, I am a bit wary at suggesting to the user that `setoid_rewrite` is a good substitute for `rewrite`. It is great in presence of setoids, morphisms, orders, etc, but when it comes to plain Leibniz' equality, its features are a plague.

Reply:
Ah OK. Note though that it is also recommended above to find occurrences to rewrite under binders.

Reply:
Which part are you referring to?

Anyway, `rewrite` has no problem finding occurrences under binders. What it cannot do is to rewrite bound occurrences (since equality is not extensional in Coq). For that, you need `setoid_rewrite`. But even so, if you just naively replace `rewrite` by `setoid_rewrite`, then you will get a crazy error message. (Just test it for yourself. That is a sight to behold. This is even worse than the error messages you get due to templates in C++.) You first have to prove (and declare) that all the functions that occur in your goal are morphisms for extensional equality. So, again, not something we should suggest the user without a lot of care.